Bugfix/do not try to migrate deleted attachments

Alexander-Hjelm opened the pull request:
Supersedes https://github.com/solidify/jira-azuredevops-migrator/pull/613

Alexander-Hjelm commented:
Integration tests run successfully: https://dev.azure.com/solidify/Internal/_build?definitionId=138&_a=summary

Revision is imported without problems:

![image](https://user-images.githubusercontent.com/10683896/202126414-1d0207cc-1f43-4fef-8b1b-805849fd8276.png)

Alexander-Hjelm commented:
Attachment count is OK

No mention of the removed attachment in the History tab:
![image](https://user-images.githubusercontent.com/10683896/202126955-30f0dc2d-7c7f-48f6-9bb5-308b1342deba.png)

madkoo commented on src/WorkItemMigrator/JiraExport/JiraItem.cs:
```diff
@@ -101,9 +101,14 @@ private static List<JiraRevision> BuildRevisions(JiraItem jiraItem, IJiraProvide
-                        attachmentChanges.Add(attachmentChange);
-
-                        UndoAttachmentChange(attachmentChange, attachments);
+                        if (UndoAttachmentChange(attachmentChange, attachments))
```
It is getting really too if-elsey, not very fond of that. Think it makes sense to move this long if-else thing in the loop to a switch Statement. Also, I would refactor out specific methods foreach step. Like HandleAttahcmentChange-> which will call out attachment-related actions

Alexander-Hjelm replied:
Resolved

madkoo commented on src/WorkItemMigrator/JiraExport/JiraItem.cs:
```diff
@@ -182,18 +187,22 @@ private static JiraRevision BuildCommentRevision(Comment c, JToken rc, JiraItem
-        private static void UndoAttachmentChange(RevisionAction<JiraAttachment> attachmentChange, List<JiraAttachment> attachments)
+        private static bool UndoAttachmentChange(RevisionAction<JiraAttachment> attachmentChange, List<JiraAttachment> attachments)
         {
+            bool result = false;
```
Same here, I would split out to specific functions instead. sure, the first if can stay but i would return false right away.
In the else i would refactor out to a method RemoveAttachment that would handle remove part and would return boolean, eg: 

if(RemovedType)
reutrn false;
return RemoveAttachmentChange(param1, param2);

Alexander-Hjelm replied:
Resolved

madkoo commented on src/WorkItemMigrator/tests/Migration.Jira-Export.Tests/JiraItemTests.cs:
```diff
@@ -24,6 +28,126 @@ public void Setup()
+            //Act
+            var jiraItem = JiraItem.CreateFromRest(issueKey, provider);
+
+            //Assert
```
Again, when using multiple asserts it is good practice to wrap it with Assert.Multiple, in that case it will run all asserts and not fail on the first on

Alexander-Hjelm replied:
Resolved

madkoo commented on src/WorkItemMigrator/tests/Migration.Jira-Export.Tests/JiraItemTests.cs:
```diff
@@ -24,6 +28,126 @@ public void Setup()
+            //Act
+            var jiraItem = JiraItem.CreateFromRest(issueKey, provider);
+
+            //Assert
```
Wrap around Assert.Multple

Alexander-Hjelm replied:
Resolved

madkoo reviewed (commented):
Check  my suggested comments, what do you think ?

Alexander-Hjelm commented:
@moerketh, please address madis's comments. You can merge into this branch from yours :)

moerketh commented:
@madkoo I think this entire region of private static methods should not be in this class. RevisionBuilding (this method) is core to this application and should be it's own class. The parsing of changes should be another and probably the handling of Links and Attachments should be separated as well.

That's a bit more work than fixing this bug ;) I'll apply your suggestions for now

moerketh commented:
@Alexander-Hjelm I've updated my branch. Do you want me to create a new PR for it?

I've also changed the loglevel for deleted attachments to "Debug" instead of "Warning". It's the expected behavior so no need to warn about deleted stuff.

Alexander-Hjelm commented:
> @Alexander-Hjelm I've updated my branch. Do you want me to create a new PR for it?
> 
> I've also changed the loglevel for deleted attachments to "Debug" instead of "Warning". It's the expected behavior so no need to warn about deleted stuff.

Sorry for the delay :) Create a new PR into this branch (bugfix/do-not-try-to-migrate-deleted-attachments) and tag me and I'll review :)

Alexander-Hjelm commented:
Integration tests: https://dev.azure.com/solidify/Internal/_build/results?buildId=9970&view=results

Alexander-Hjelm commented:
Integration tests (latest): https://dev.azure.com/solidify/Internal/_build/results?buildId=11071&view=results
